bigframes/core/__init__.py

```diff
@@ -49,470 +40,196 @@
-        """Creates a mutable builder for expressions."""
-        # Since ArrayValue is intended to be immutable (immutability offers
-        # potential opportunities for caching, though we might need to introduce
-        # more node types for that to be useful), we create a builder class.
-        return ArrayValueBuilder(
-            self._session,
-            self._table,
-            columns=self._columns,
-            hidden_ordering_columns=self._hidden_ordering_columns,
-            ordering=self._ordering,
-            predicates=self._predicates,
+    def cached(self, cluster_cols: typing.Sequence[str]) -> ArrayValue:
+        """Write the ArrayValue to a session table and create a new block object that references it."""
+        compiled = self.compile()
+        ibis_expr = compiled._to_ibis_expr("unordered", expose_hidden_cols=True)
+        destination = self.session._ibis_to_session_table(
+            ibis_expr, cluster_cols=cluster_cols, api_name="cache"
         )
-
-    def drop_columns(self, columns: Iterable[str]) -> ArrayValue:
-        # Must generate offsets if we are dropping a column that ordering depends on
-        expr = self
-        for ordering_column in set(columns).intersection(
-            [col.column_id for col in self._ordering.ordering_value_columns]
-        ):
-            expr = self._hide_column(ordering_column)
-
-        expr_builder = expr.builder()
-        remain_cols = [
-            column for column in expr.columns if column.get_name() not in columns
-        ]
-        expr_builder.columns = remain_cols
-        return expr_builder.build()
-
-    def get_column_type(self, key: str) -> bigframes.dtypes.Dtype:
-        ibis_type = typing.cast(
-            bigframes.dtypes.IbisDtype, self._get_any_column(key).type()
+        table_expression = self.session.ibis_client.table(
+            f"{destination.project}.{destination.dataset_id}.{destination.table_id}"
         )
-        return typing.cast(
-            bigframes.dtypes.Dtype,
-            bigframes.dtypes.ibis_dtype_to_bigframes_dtype(ibis_type),
+        new_columns = [table_expression[column] for column in compiled.column_ids]
+        new_hidden_columns = [
+            table_expression[column]
+            for column in compiled._hidden_ordering_column_names
+        ]
+        return ArrayValue.from_ibis(
+            self.session,
+            table_expression,
+            columns=new_columns,
+            hidden_ordering_columns=new_hidden_columns,
+            ordering=compiled._ordering,
         )
 
-    def _get_ibis_column(self, key: str) -> ibis_types.Value:
-        """Gets the Ibis expression for a given column."""
-        if key not in self.column_ids:
-            raise ValueError(
-                "Column name {} not in set of values: {}".format(key, self.column_ids)
-            )
-        return typing.cast(ibis_types.Value, self._column_names[key])
+    def _get_table_size(self, destination_table):
```

Comment:
Who calls this? Maybe we can have a separate refactor PR to delete `_get_table_size()` from ArrayValue before we do this PR.

Reply:
inlined this